File: files.py
import os

import pandas as pd
from timeit import default_timer as timer

import util


# File Reading ---------------------------------------------------------------------------------------------------------

def to_messages(path):
    temp = timer()
    txt = open(path, 'r', encoding="utf8")
    messages = []
    lines = txt.readlines()
    txt.close()
    count = len(lines)
    i = -1
    for index in range(count):
        try:
            line = lines[index]
        except IndexError:
            break
        if util.date(line[0:10]):
            i += 1
            # first line in message
            messages.append(line)
        else:
            # non-initial line in message
            messages[i] += line
    util.append_time(timer() - temp)
    return messages


def get_dfs(name):
    try:
        return (pd.read_csv(f'saved/{name}/chat_messages.csv', header=0, index_col='index'),
                pd.read_csv(f'saved/{name}/chat_time.csv', header=0, index_col='index'),
                pd.read_csv(f'saved/{name}/chat_characters.csv', header=0, index_col='char'),
                pd.read_csv(f'saved/{name}/chat_message_types.csv', header=0, index_col='index'),
                pd.read_csv(f'saved/{name}/chat_words.csv', header=0, index_col='index'),
                pd.read_csv(f'saved/{name}/chat_ioc.csv', header=0, index_col='index'),
                pd.read_csv(f'saved/{name}/chat_blocks.csv', header=0, index_col='index'))
    except FileNotFoundError:
        print(f"`saved/{name}/chat_....csv` were not found. Try to generate the Database again.")
        return (None,) * 7


def get_df(name):
    try:
        return pd.read_csv(f'saved/{name}/chat.csv', header=0)
    except FileNotFoundError:
        print(f"`saved/{name}/chat.csv` not found. Try to generate the Database again.")
        return None


# File Saving ----------------------------------------------------------------------------------------------------------

def save_df(df, name):
    folder(name)
    df.index.name = 'index'
    df.to_csv(f'saved/{name}/chat.csv', encoding="utf-8")


def save_dfs(mdf, tdf, cdf, mtdf, wdf, idf, bdf, name):
    folder(name)
    mdf.index.name = 'index'
    tdf.index.name = 'index'
    if cdf is not None:
        cdf.index.name = 'char'
    mtdf.index.name = 'index'
    if wdf is not None:
        wdf.index.name = 'index'
    idf.index.name = 'index'
    bdf.index.name = 'index'
    mdf.to_csv(f'saved/{name}/chat_messages.csv', encoding="utf-8")
    tdf.to_csv(f'saved/{name}/chat_time.csv', encoding="utf-8")
    if cdf is not None:
        cdf.to_csv(f'saved/{name}/chat_characters.csv', encoding="utf-8")
    mtdf.to_csv(f'saved/{name}/chat_message_types.csv', encoding="utf-8")
    if wdf is not None:
        wdf.to_csv(f'saved/{name}/chat_words.csv', encoding="utf-8")
    idf.to_csv(f'saved/{name}/chat_ioc.csv', encoding="utf-8")
    bdf.to_csv(f'saved/{name}/chat_blocks.csv', encoding="utf-8")


def folder(name):
    if not os.path.exists(f"saved/{name}"):
        os.makedirs(f"saved/{name}")

File: Files.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;

public class Files {
    public static String[] toMessages(String[] lines) {
        ArrayList<String> messages = new ArrayList<>();
        int i = -1;
        for (String line : lines) {
            if (Util.isDate(line)) {
                i += 1;
                // first line in message
                messages.add(line);
            } else {
                // non-initial line in message
                messages.set(i, messages.get(i) + "\n" + line);
            }
        }
        return messages.toArray(new String[0]);
    }

    public static void saveDF(NormalDataFrame df, String name) {
        folder("saved/" + name);
        df.index("index");
        df.toCSV("saved/" + name + "/chat.csv");
    }

    public static NormalDataFrame getDF(String name) {
        try {
            return DataFrame.readCSV("saved/" + name + "/chat.csv");
        }
        catch (FileNotFoundException e) {
            Util.print("`saved/" + name + "/chat.csv` not found. Try to generate the Database again.");
            return null;
        }
    }

    private static void folder(String name) {
        File dir = new File(name);
        if (!dir.exists()) dir.mkdirs();
    }

    public static void saveDFs(AlphabetsDataFrame adf, TimeDataFrame tdf, CharacterDataFrame cdf, MessageTypeDataFrame mtdf, WordsDataFrame wdf, IOCDataFrame idf, BlocksDataFrame bdf, String name) {
        folder("saved/" + name);
        if (adf != null) adf.index("index");
        if (tdf != null) tdf.index("index");
        if (cdf != null) cdf.index("char");
        if (mtdf != null) mtdf.index("index");
        if (wdf != null) wdf.index("index");
        if (idf != null) idf.index("index");
        if (bdf != null) bdf.index("index");

        if (adf != null) adf.toCSV("saved/" + name + "/chat_alphabets.csv");
        if (tdf != null) tdf.toCSV("saved/" + name + "/chat_time.csv");
        if (cdf != null) cdf.toCSV("saved/" + name + "/chat_characters.csv");
        if (mtdf != null) mtdf.toCSV("saved/" + name + "/chat_message_types.csv");
        if (wdf != null) wdf.toCSV("saved/" + name + "/chat_words.csv");
        if (idf != null) idf.toCSV("saved/" + name + "/chat_ioc.csv");
        if (bdf != null) bdf.toCSV("saved/" + name + "/chat_blocks.csv");
    }
}
